Feature comma dangle allow multiline

keithamus opened the pull request:
Fixes #1984 

Adds option for `"allow-multiline"` to the comma-dangle rule, which fails if single-line objects/arrays have a dangling comma, or multiline objects/arrays are missing a dangling comma.

keithamus commented:
Failures are unrelated to the actual fix.

nzakas commented on lib/rules/comma-dangle.js:
```diff
@@ -12,7 +12,9 @@
 
 module.exports = function (context) {
-    var forbidDanglingComma = context.options[0] !== "always";
+    var allowDangle = context.options[0] || "never";
```
You're changing the default behavior here, which isn't necessary to add the new option. Please revert.

UltCombo replied:
@nzakas is it so? Looks like the default (if the option argument is undefined) is still `never`.

nzakas replied:
Look at what happens when `context.options[0]` is `"foo"`:

* `forbidDanglingComma` is `true` (equivalent to `"never"`)
* `allowDangle` is `"foo"` (equivalent to `"always"` based on logic around line 38)

UltCombo replied:
I'm not sure what the ESLint convention is here, but wouldn't it make sense to throw for unrecognized options? Throwing errors makes it easier for the user to spot errors such as typos.

nzakas replied:
This is our convention: use a logical default. Rules aren't allow to throw.

There's a separate discussion around how rules should be validated here: #967

UltCombo replied:
Oh I see, thanks for the link.

nzakas commented on lib/rules/comma-dangle.js:
```diff
@@ -12,7 +12,9 @@
 module.exports = function (context) {
-    var forbidDanglingComma = context.options[0] !== "always";
+    var allowDangle = context.options[0] || "never";
+    var unexpectedMessage = "Unexpected trailing comma.";
```
These should be in UPPER_CASE since they don't change.

nzakas commented:
Yeah, master is failing thanks to escope upgrade. I'm working on fixing that. In the meantime, please read over our contribution guidelines to see how to properly format your pull request: http://eslint.org/docs/developer-guide/contributing.html

keithamus commented:
@nzakas I've rebased and squashed into one - which is why I presume you pointed me to the contributing guide. If not could you please explain the exact points I've missed.

I've fixed the behaviour you have specified in the above comments - and added tests to ensure that when called with erroneous strings the behaviour defaults to "never". 

Let me know if there is anything else. Thanks.

gabrielecirulli commented:
I would really like this! :+1:

gabrielecirulli commented:
Is there any ETA? (just to get an idea)

nzakas commented:
@keithamus there are no failures on master right now, and your PR is also not failing. If it's failing locally, be sure to `npm install` to get the latest dependencies.

Have you also signed the CLA?

keithamus commented:
@nzakas CLA is signed! I signed it before I submitted the PR. Can I suggest using https://www.clahub.com/ if you're having difficulty keeping track of CLA signing - as it'll use the Github status API to warn you of committers who haven't signed the CLA.

Also this PR is now passing. Travis is passing. As far as I know this PR is good to go. If you have any more notes please let me know!

nzakas commented:
I'm already working on a bot to do the same thing, but thanks for the suggestion. I usually just check the GDoc spreadsheet but was on mobile when I responded last.

nzakas commented on docs/rules/comma-dangle.md:
```diff
@@ -17,6 +17,8 @@ This rule is aimed to forbid or enforce trailing commas in object literals and a
 
 This rule takes one argument. If it is `"always"` then it warns whenever a missing comma is detected.
 If `"never"` then it warns whenever an trailing comma is detected.
+If `"always-multiline"` then it warns if there is a missing trailing comma on arrays or objects that
```
Can you move this above the description of never?

keithamus commented:
Updated.

nzakas commented:
Thanks. I'm going to hold off on merging because I might do one more patch release, but this is good to go.

keithamus commented:
Ooh @nzakas you tease!

keithamus commented:
:tada:
